project_recalculate/models/project_project.py

```diff
@@ -0,0 +1,100 @@
+                      if self.date_start
+                      else fields.Datetime.from_string(self.date))
+        date_end = (fields.Datetime.from_string(self.date)
+                    if self.date
+                    else fields.Datetime.from_string(self.date_start))
+        if len(project_task_ids) > 0:
+            min_date_start = fields.Datetime.from_string(
+                project_task_ids[0].date_start)
+            max_date_start = fields.Datetime.from_string(
+                project_task_ids[0].date_start)
+            max_date_end = fields.Datetime.from_string(
+                project_task_ids[0].date_end)
+            date_start = fields.Datetime.from_string(
+                project_task_ids[0].date_start)
+            for project_task in project_task_ids:
+                if not project_task.date_start and project_task.date_end:
+                    project_task.date_start = project_task.date_end
+                if not project_task.date_end and project_task.date_start:
+                    project_task.date_end = project_task.date_start
+                if not project_task.date_start or not project_task.date_end:
+                    continue
+                date_start = fields.Datetime.from_string(
+                    project_task.date_start)
+                date_end = fields.Datetime.from_string(project_task.date_end)
+                if min_date_start > date_start:
+                    min_date_start = date_start
+                if max_date_end < date_end:
+                    max_date_end = date_end
+                if max_date_start < date_start:
+                    max_date_start = date_end
+            if self.calculation_type == 'date_begin':
+                vals['date'] = max_date_end
+            else:
+                vals['date_start'] = min_date_start
+        self.write(vals)
+
+    def project_recalculate(self):
+        if not self.calculation_type:
+            raise Warning(_('Cannot recalculate project because your project '
+                            'don\'t have calculation type.'))
+        if self.calculation_type == 'date_begin' and not self.date_start:
+            raise Warning(_('Cannot recalculate project because your project '
+                            'don\'t have date start.'))
+        if self.calculation_type == 'date_end' and not self.date:
+            raise Warning(_('Cannot recalculate project because your project '
+                            'don\'t have date end.'))
+        project_task_obj = self.env['project.task']
+        project_task_type_obj = self.env['project.task.type']
+        project_task_type = project_task_type_obj.search([('fold', '=', True)])
+        project_task_type_ids = [x.id for x in project_task_type]
+        project_task_ids = project_task_obj.search(
+            [('project_id', '=', self.id),
+             ('stage_id', 'not in', project_task_type_ids)
+             ])
+        for project_task in project_task_ids:
```

Comment:
I suggest:

    for task in self.task_ids:
        if not task.stage_id.fold:
            ...